### PR TITLE
Add pre/psycho/history to Part

### PR DESCRIPTION
**Pull request by pleroy**
This is the second part of #1513.

### CODE REVIEW

File: ksp_plugin/part.hpp [hunk 1]
```diff
@@ -118,8 +133,21 @@ class Part final {
       containing_pile_up_;
 
   DegreesOfFreedom<Barycentric> degrees_of_freedom_;
-  not_null<std::unique_ptr<DiscreteTrajectory<Barycentric>>> tail_;
-  bool tail_is_authoritative_ = false;
+
+  // See the comments in pile_up.hpp for an explanation of the terminology.
+
+  // The |prehistory_| always has a single point at time -∞.  It sole purpose is
+  // to make it convenient to hook the |psychohistory_| even if these is no
```
Comment (eggrobin) [inline ksp_plugin/part.hpp]: there

File: ksp_plugin/part.hpp [hunk 2]
```diff
@@ -118,8 +133,21 @@ class Part final {
...
+  // point in the |history_| (it's not possible to fork-at-last an empty root
+  // trajectory, but it works for a non-root).
+  not_null<std::unique_ptr<DiscreteTrajectory<Barycentric>>> prehistory_;
+  // The |history_| is nearly always not null, except in some transient
+  // situations.  It's a fork of the |prehistory_|.
+  DiscreteTrajectory<Barycentric>* history_ = nullptr;
+  // The |psychohistory_| is destroyed by |AppendToHistory| and is recreated
+  // as needed by |AppendToPsychohistory| or by |tail|.  That's because
+  // |FortAtLast| is relatively expensive so we only call it when necessary.
```
Comment (eggrobin) [inline ksp_plugin/part.hpp]: Fork

File: ksp_plugin/part.hpp [hunk 3]
```diff
@@ -74,12 +75,26 @@ class Part final {
   // to |PileUp::AdvanceTime| for the containing |PileUp|.  It is read and
   // cleared by |Vessel::AdvanceTime| for the containing |Vessel|.
   DiscreteTrajectory<Barycentric>& tail();
-  DiscreteTrajectory<Barycentric> const& tail() const;
 
   // True if and only if the last point of the tail is authoritative, i.e.,
-  // corresponds to a point in the psychohistory of the enclosing Part.
+  // corresponds to a point in the psychohistory of the enclosing |Vessel|.
```
Comment (eggrobin) [inline ksp_plugin/part.hpp]: I'm confused. Should that be "a point in the history"? what is "tail()" at this point anyway?

Comment (pleroy) [inline ksp_plugin/part.hpp]: Changed to "history" but note that I didn't really change that comment and it didn't make sense from day 1.

File: ksp_plugin/part.hpp (hunk 1 above)
Comment (pleroy) [inline ksp_plugin/part.hpp]: Done.

File: ksp_plugin/part.hpp (hunk 2 above)
Comment (pleroy) [inline ksp_plugin/part.hpp]: Done.

**Pull request merged**